### PR TITLE
Refactor:  Separate Pipeline Management into Serverless Application

### PR DESCRIPTION
**Pull request by StewartW**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/deployment_map.py [hunk 1]
```diff
@@ -101,4 +100,4 @@ def _process_dir(self, path):
                     self._read(filename)
                 )
             else:
-                LOGGER.warning("%s is not a directory and doesn't end in.yml", filename)
+                LOGGER.warning(f"{filename} is not a directory and doesn't end in.yml")
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/deployment_map.py:103]: Could you please revert this change? The logger has a capability to insert strings only when the logging level is enabled. So the filename will not be inserted if logging is disabled. In this case probably all loggers would pick up on warnings. However, I would like to keep those consistent if possible.

```suggestion
                LOGGER.warning("%s is not a directory and doesn't hold the .yml suffix", filename)
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/repo.py [hunk 2]
```diff
@@ -23,8 +22,6 @@
     DEPLOYMENT_ACCOUNT_REGION,
     S3_BUCKET_NAME
 )
-
-
 class Repo:
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/repo.py:25]: The lines that were removed before the class should be there.
Adding these as a suggestion here, as GitHub doesn't support suggestions on deleted lines (odd).
```suggestion


class Repo:
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/rule.py [hunk 3]
```diff
@@ -54,5 +54,5 @@ def create_update(self):
             s3_key_path=None,
             account_id=DEPLOYMENT_ACCOUNT_ID,
         )
-        LOGGER.info('Ensuring Stack State for Event Rule forwarding from %s to %s', self.source_account_id, DEPLOYMENT_ACCOUNT_ID)
+        LOGGER.info(f'Ensuring Stack State for Event Rule forwarding from {self.source_account_id} to {DEPLOYMENT_ACCOUNT_ID}')
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/rule.py:57]: Same comment as above on the logger and use of `%s` being preferred over formatted strings with loggers:
```suggestion
        LOGGER.info('Ensuring Stack State for Event Rule forwarding from %s to %s', self.source_account_id, DEPLOYMENT_ACCOUNT_ID)
```

File: src/template.yml [hunk 4]
```diff
@@ -17,12 +17,12 @@ Metadata:
     Labels:
       ["adf", "aws-deployment-framework", "multi-account", "cicd", "devops"]
     HomePageUrl: https://github.com/awslabs/aws-deployment-framework
-    SemanticVersion: 3.1.2
+    SemanticVersion: 3.4.1
```
Comment (sbkok) [inline src/template.yml:20]: This should not be changed in this PR:
```suggestion
    SemanticVersion: 3.1.2
```

File: src/template.yml [hunk 5]
```diff
@@ -17,12 +17,12 @@ Metadata:
...
     SourceCodeUrl: https://github.com/awslabs/aws-deployment-framework
 Mappings:
   Metadata:
     ADF:
-      Version: 3.1.2
+      Version: 3.4.1
```
Comment (sbkok) [inline src/template.yml:25]: This should not be changed in this PR:

```suggestion
      Version: 3.1.2
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/target.py [hunk 6]
```diff
@@ -48,13 +48,13 @@ def _define_target_type(target):
...
     def generate_waves(self):
-        waves = []
         wave_size = self.wave.get('size', 50)
+        wave = []
         length = len(self.account_list)
         for index in range(0, length, wave_size):
-            yield self.account_list[index:min(index + wave_size, length)]
-            waves.append(self.account_list[index:min(index + wave_size, length)])
-        return waves
+            wave.append(self.account_list[index:min(index + wave_size, length)])
```
Comment (StewartW) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/target.py:55]: I've changed this back as it had issues serializing the response due to a generator being used.

File: src/template.yml (hunk 4 above)
Comment (StewartW) [reply src/template.yml:20]: Woops, changed this for deploying locally

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/cloudformation.py [hunk 7]
```diff
@@ -178,21 +180,21 @@ def _create_change_set(self):
...
-                    ChangeSetType=self._get_change_set_type())
+                change_set_params = {
+                    "StackName": self.stack_name,
+                    "TemplateURL": self.template_url,
+                    "Parameters": self.parameters if self.parameters is not None else self.get_parameters(),
+                    "Capabilities": ["CAPABILITY_NAMED_IAM", "CAPABILITY_AUTO_EXPAND"],
+                    "Tags":[{
+                        'Key': 'createdBy',
+                        'Value': 'ADF'
+                        }],
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/cloudformation.py:191]:
```suggestion
                    }],
```

**Review by sbkok — CHANGES_REQUESTED** [review]

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 8]
```diff
@@ -0,0 +1,879 @@
...
+                  - cloudformation:ExecuteChangeSet
+                  - cloudformation:SetStackPolicy
+                  - cloudformation:ValidateTemplate
+                Resource:
+                  - "*"
+              - Effect: Allow
+                Sid: "PassRole"
+                Action:
+                  - 'iam:PassRole'
+                Resource: '*'
```
[the remark this answers is not in the document]
Comment (StewartW) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:607]: I think that can be removed. (It's a carry over.) Can you just double check. I think the CodeBuild Project role will never need to  pass an IAM role

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 9]
```diff
@@ -0,0 +1,879 @@
...
+                  - "iam:CreateRole"
+                  - "iam:CreateRolePolicy"
+                  - "iam:DeleteRole"
+                  - "iam:DeleteRolePolicy"
+                  - "iam:GetRole"
+                  - "iam:GetRolePolicy"
+                  - "iam:PassRole"
+                  - "iam:PutRolePolicy"
+                Resource:
+                  - !Sub arn:${AWS::Partition}:iam::${AWS::AccountId}:role/*
```
[the remark this answers is not in the document]
Comment (StewartW) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:676]: I think this can be removed as well. Based on my reasoning this role should only be used to deploy CodePipeline stacks which don't include roles. 

Think this is just hangover from previous iterations.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 10]
```diff
@@ -0,0 +1,886 @@
...
+      Description: "ADF Lambda Function - Create Repository"
+      Environment:
+        Variables:
+          ACCOUNT_ID: !Ref AWS::AccountId
+          ORGANIZATION_ID: !Ref OrganizationID
+          ADF_VERSION: !Ref ADFVersion
+          ADF_LOG_LEVEL: !Ref ADFLogLevel
+          ADF_ROLE_NAME: !Ref CrossAccountAccessRole
+          S3_BUCKET_NAME: !Ref PipelineBucket
+      FunctionName: CreateRepositoryFunction
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:777]:
```suggestion
      FunctionName: ADFPipelineCreateRepositoryFunction
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 11]
```diff
@@ -0,0 +1,886 @@
...
+      Environment:
+        Variables:
+          ACCOUNT_ID: !Ref AWS::AccountId
+          ORGANIZATION_ID: !Ref OrganizationID
+          ADF_VERSION: !Ref ADFVersion
+          ADF_LOG_LEVEL: !Ref ADFLogLevel
+          ADF_ROLE_NAME: !Ref CrossAccountAccessRole
+          S3_BUCKET_NAME: !Ref PipelineBucket
+          ROOT_ACCOUNT_ID: !Ref RootAccountId
+      FunctionName: GeneratePipelineInputsFunction
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:794]:
```suggestion
      FunctionName: ADFPipelineGenerateInputsFunction
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 12]
```diff
@@ -0,0 +1,886 @@
...
+      Environment:
+        Variables:
+          ACCOUNT_ID: !Ref AWS::AccountId
+          ORGANIZATION_ID: !Ref OrganizationID
+          ADF_VERSION: !Ref ADFVersion
+          ADF_LOG_LEVEL: !Ref ADFLogLevel
+          ADF_ROLE_NAME: !Ref CrossAccountAccessRole
+          S3_BUCKET_NAME: !Ref ADFDefinitionBucket
+          ROOT_ACCOUNT_ID: !Ref RootAccountId
+      FunctionName: StoreDefinitionFunction
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:811]:
```suggestion
      FunctionName: ADFPipelineStoreDefinitionFunction
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 13]
```diff
@@ -0,0 +1,886 @@
...
+        Variables:
+          ACCOUNT_ID: !Ref AWS::AccountId
+          ORGANIZATION_ID: !Ref OrganizationID
+          ADF_VERSION: !Ref ADFVersion
+          ADF_LOG_LEVEL: !Ref ADFLogLevel
+          ADF_ROLE_NAME: !Ref CrossAccountAccessRole
+          ROOT_ACCOUNT_ID: !Ref RootAccountId
+          S3_BUCKET_NAME: !Ref ADFPipelineBucket
+          ADF_PIPELINE_PREFIX: !Ref PipelinePrefix
+      FunctionName: IdentifyOutOfDatePipelinesFunction
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:829]:
```suggestion
      FunctionName: ADFPipelineIdentifyOutOfDatePipelinesFunction
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/deployment_map.py [hunk 14]
```diff
@@ -83,10 +92,8 @@ def _get_all(self):
         )
         if not self.map_contents['pipelines']:
             LOGGER.error(
-                "No Deployment Map files found, create a deployment_map.yml "
-                "file in the root of the repository to create pipelines. "
-                "You can create additional deployment maps if required in a "
-                "folder named deployment_maps with any name (ending in .yml)"
+                "No Deployment Map files found, create a deployment_map.yml file in the root of the repository to create pipelines. "
+                "You can create additional deployment maps if required in a folder named deployment_maps with any name (ending in .yml)"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/deployment_map.py:96]: Can you revert this change, so the line length matches what we defined in the .editorconfig file?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/tests/test_deployment_map.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+                        "regions": ["eu-west-1"],
+                        "step_name": "",
+                    },
+                ]
+            ]
+        ]
+    }
+
+    cls.update_deployment_parameters(pipeline)
+    assert cls.account_ou_names["some_account"] == "/fake/path"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-build/shared/python/tests/test_deployment_map.py:68]: Could you add the tests I added in the #484 Fix: Wave deployments PR. That way I can close that one in favour of this one :smile:

See https://github.com/awslabs/aws-deployment-framework/pull/484/commits/18a0492b46ade867baa8f1e7256d72bbbf92f93b

**Review by sbkok — CHANGES_REQUESTED** [review]
Great improvements, thanks for making those.
I've got a number of minor suggested changes, so we are close to merge this too!

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (sbkok) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml]: I think it uses the PassRole when executing the CloudFormation stacks.
According to the change you introduced, it will always use one role, see https://github.com/awslabs/aws-deployment-framework/pull/424/files/7f9d1c95cba423d9e20533436652b6d172399ed3#diff-de3040c3b70530a31a46f3a9ce5ffde48ecf03da5153a64929720b8e53c334e6R58

The role is created and configured in this stack too, see https://github.com/awslabs/aws-deployment-framework/pull/424/files/7f9d1c95cba423d9e20533436652b6d172399ed3#diff-dead502985c4f4a35238ba4a93fcedfabd8d013bd70e545003066d4a01afe956R500-R501

So I think we can restrict it to:

```suggestion
                Resource:
                  - !GetAtt ADFPipelineMangementCloudFormationRole.Arn
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (sbkok) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml]: Unfortunately it does create a role depending on the resources in the pipeline.
For example, for CodePipeline it needs to create a role that has access.

That role name starts with `adf-pipeline-`. So maybe we can restrict this to only accept that as the prefix. What do you think?

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 16]
```diff
@@ -0,0 +1,886 @@
...
+                  - "sns:Subscribe"
+                  - "sns:SetTopicAttributes"
+                  - "sns:GetTopicAttributes"
+                  - "sns:TagResource"
+                Resource:
+                  - !Sub arn:${AWS::Partition}:sns:${AWS::Region}:${AWS::AccountId}:${PipelinePrefix}*
+              - Effect: Allow
+                Action:
+                  - "iam:CreateRole"
+                  - "iam:CreateRolePolicy"
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:672]: This action does not exist, do you mean AttachRolePolicy?

```suggestion
                  - "iam:AttachRolePolicy"
```

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 17]
```diff
@@ -0,0 +1,886 @@
...
+                  - "iam:DeleteRole"
+                  - "iam:DeleteRolePolicy"
+                  - "iam:GetRole"
+                  - "iam:GetRolePolicy"
+                  - "iam:PutRolePolicy"
+                Resource:
+                  - !Sub arn:${AWS::Partition}:iam::${AWS::AccountId}:role/aws-deployment-framework/*
```
[the remark this answers is not in the document]
Comment (sbkok) [reply src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:679]: Maybe we can have a separate scope for pipeline roles. So a pipeline will not be granted access to the more permissive admin roles that ADF needs to generate resources. Or we define a separate path for those.

File: src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml [hunk 18]
```diff
@@ -0,0 +1,886 @@
...
+                Condition:
+                  StringEquals:
+                    aws:PrincipalOrgID: !Ref OrganizationID
```
Comment (sbkok) [inline src/lambda_codebase/initial_commit/bootstrap_repository/adf-bootstrap/deployment/pipeline_management.yml:682]: This condition is not supported on any of the IAM actions listed above.
This statement would work with PassRole, and we still need that I am afraid.
As CodePipeline, for example, gets passed the role that we create using the allowed actions defined in this statement.

**Review by sbkok — CHANGES_REQUESTED** [review]

**Review by sbkok — APPROVED** [review]
Thanks for making all these changes @StewartW!
This will help speed up ADF a lot, excellent work!

**Review by javydekoning — APPROVED** [review]
This looks good to me! 

I suggest also storing the `.drawio` originals to the repo so the diagram can be easily updated when there are changes to the statemachine later.

**Comment by sbkok** [comment]
For reference, the .drawio.png files that were added by the PR also include the source of the image.
You can edit them by opening the png file directly with the DrawIO tool.
